### PR TITLE
13.0 mig account budget oca

### PR DESCRIPTION
**Pull request by amcor**
cc @guadaltech

### CODE REVIEW

File: account_budget_oca/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+{
+    "name": "Budgets Management",
+    "version": "13.0.1.0.0",
+    "category": "Accounting",
+    "license": "LGPL-3",
+    "author": "Odoo S.A., " "Odoo Community Association (OCA)",
```
Comment (astirpe) [inline account_budget_oca/__manifest__.py]:
```suggestion
    "author": "Odoo S.A., Odoo Community Association (OCA)",
```

File: account_budget_oca/security/account_budget_security.xml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <data noupdate="1">
+        <record id="budget_post_comp_rule" model="ir.rule">
+            <field name="name">Budget post multi-company</field>
+            <field name="model_id" ref="model_account_budget_post" />
+            <field eval="True" name="global" />
+            <field
+                name="domain_force"
+            >['|',('company_id','=',False),('company_id','child_of',[user.company_id.id])]</field>
```
Comment (astirpe) [inline account_budget_oca/security/account_budget_security.xml]:
```suggestion
            >['|',('company_id','=',False),('company_id','in',company_ids))]</field>
```

File: account_budget_oca/security/account_budget_security.xml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        </record>
+        <record id="budget_comp_rule" model="ir.rule">
+            <field name="name">Budget multi-company</field>
+            <field name="model_id" ref="model_crossovered_budget" />
+            <field eval="True" name="global" />
+            <field
+                name="domain_force"
+            >['|',('company_id','=',False),('company_id','child_of',[user.company_id.id])]</field>
```
Comment (astirpe) [inline account_budget_oca/security/account_budget_security.xml]:
```suggestion
            >['|',('company_id','=',False),('company_id','in',company_ids))]</field>
```

File: account_budget_oca/security/account_budget_security.xml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        </record>
+        <record id="budget_lines_comp_rule" model="ir.rule">
+            <field name="name">Budget lines multi-company</field>
+            <field name="model_id" ref="model_crossovered_budget_lines" />
+            <field eval="True" name="global" />
+            <field
+                name="domain_force"
+            >['|',('company_id','=',False),('company_id','child_of',[user.company_id.id])]</field>
```
Comment (astirpe) [inline account_budget_oca/security/account_budget_security.xml]:
```suggestion
            >['|',('company_id','=',False),('company_id','in',company_ids))]</field>
```

File: account_budget_oca/models/account_budget.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    _description = "Budgetary Position"
+
+    name = fields.Char(required=True)
+    account_ids = fields.Many2many(
+        comodel_name="account.account",
+        relation="account_budget_rel",
+        column1="budget_id",
+        column2="account_id",
+        string="Accounts",
+        domain=[("deprecated", "=", False)],
```
Comment (astirpe) [inline account_budget_oca/models/account_budget.py]:
```suggestion
        domain="[('company_id', '=', company_id), ('deprecated', '=', False)]",
```

File: account_budget_oca/models/account_budget.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    crossovered_budget_line_ids = fields.One2many(
+        comodel_name="crossovered.budget.lines",
+        inverse_name="general_budget_id",
+        string="Budget Lines",
+    )
+    company_id = fields.Many2one(
+        comodel_name="res.company",
+        string="Company",
+        required=True,
+        default=lambda self: self.env["res.company"]._company_default_get(
```
Comment (astirpe) [inline account_budget_oca/models/account_budget.py]:
```python
default=lambda self: self.env.company,
```

File: account_budget_oca/models/account_budget.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        domain=[("deprecated", "=", False)],
+    )
+    crossovered_budget_line_ids = fields.One2many(
+        comodel_name="crossovered.budget.lines",
+        inverse_name="general_budget_id",
+        string="Budget Lines",
+    )
+    company_id = fields.Many2one(
+        comodel_name="res.company",
+        string="Company",
```
Comment (astirpe) [inline account_budget_oca/models/account_budget.py]:
```suggestion
```

File: account_budget_oca/models/account_budget.py [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+            )
+        else:
+            account_ids = self.account_ids
+        if not account_ids:
+            raise ValidationError(_("The budget must have at least one account."))
+
+    @api.model
+    def create(self, vals):
+        self._check_account_ids(vals)
+        return super(AccountBudgetPost, self).create(vals)
```
Comment (astirpe) [inline account_budget_oca/models/account_budget.py]:
```suggestion
        return super().create(vals)
```

File: account_budget_oca/models/account_budget.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+
+    def write(self, vals):
+        self._check_account_ids(vals)
+        return super(AccountBudgetPost, self).write(vals)
```
Comment (astirpe) [inline account_budget_oca/models/account_budget.py]:
```suggestion
        return super().write(vals)
```

File: account_budget_oca/models/account_budget.py [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+    crossovered_budget_line_ids = fields.One2many(
+        comodel_name="crossovered.budget.lines",
+        inverse_name="crossovered_budget_id",
+        string="Budget Lines",
+        states={"done": [("readonly", True)]},
+        copy=True,
+    )
+    company_id = fields.Many2one(
+        comodel_name="res.company",
+        string="Company",
```
Comment (astirpe) [inline account_budget_oca/models/account_budget.py]:
```suggestion
```

File: account_budget_oca/models/account_budget.py [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+        required=True,
+        default=lambda self: self.env["res.company"]._company_default_get(
```
Comment (astirpe) [inline account_budget_oca/models/account_budget.py]:
```python
default=lambda self: self.env.company,
```

File: account_budget_oca/models/account_budget.py [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+
+    crossovered_budget_id = fields.Many2one(
+        comodel_name="crossovered.budget",
+        string="Budget",
+        ondelete="cascade",
+        index=True,
+        required=True,
+    )
+    analytic_account_id = fields.Many2one(
+        comodel_name="account.analytic.account", string="Analytic Account"
```
Comment (astirpe) [inline account_budget_oca/models/account_budget.py]:
```suggestion
        comodel_name="account.analytic.account",
```

File: account_budget_oca/models/account_budget.py [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+    date_to = fields.Date(string="End Date", required=True)
+    paid_date = fields.Date()
+    planned_amount = fields.Float(required=True, digits=0)
+    practical_amount = fields.Float(compute="_compute_practical_amount", digits=0)
+    theoretical_amount = fields.Float(compute="_compute_theoretical_amount", digits=0)
+    percentage = fields.Float(compute="_compute_percentage", string="Achievement")
+    company_id = fields.Many2one(
+        related="crossovered_budget_id.company_id",
+        comodel_name="res.company",
+        string="Company",
```
Comment (astirpe) [inline account_budget_oca/models/account_budget.py]:
```suggestion
```

File: account_budget_oca/models/account_budget.py [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+    date_from = fields.Date(string="Start Date", required=True)
+    date_to = fields.Date(string="End Date", required=True)
+    paid_date = fields.Date()
+    planned_amount = fields.Float(required=True, digits=0)
+    practical_amount = fields.Float(compute="_compute_practical_amount", digits=0)
+    theoretical_amount = fields.Float(compute="_compute_theoretical_amount", digits=0)
+    percentage = fields.Float(compute="_compute_percentage", string="Achievement")
+    company_id = fields.Many2one(
+        related="crossovered_budget_id.company_id",
+        comodel_name="res.company",
```
Comment (astirpe) [inline account_budget_oca/models/account_budget.py]:
```suggestion
```

File: account_budget_oca/models/account_budget.py [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+        string="Company",
+        store=True,
+        readonly=True,
+    )
+
+    def _compute_practical_amount(self):
```
Comment (astirpe) [inline account_budget_oca/models/account_budget.py:154]:
```suggestion
    @api.depends('general_budget_id.account_ids', 'date_from', 'date_to', 'analytic_account_id')
    def _compute_practical_amount(self):
```

File: account_budget_oca/models/account_budget.py [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+                    WHERE account_id=%s
+                        AND (date between %s
+                        AND %s)
+                        AND general_account_id=ANY(%s)""",
+                    (line.analytic_account_id.id, date_from, date_to, acc_ids,),
+                )
+                result = self.env.cr.fetchone()[0] or 0.0
+            line.practical_amount = result
+
+    def _compute_theoretical_amount(self):
```
Comment (astirpe) [inline account_budget_oca/models/account_budget.py:175]:
```suggestion
    @api.depends('paid_date', 'date_from', 'date_to', 'planned_amount')
    def _compute_theoretical_amount(self):
```

File: account_budget_oca/models/account_budget.py [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+            # Used for the report
+
+            if line.paid_date:
+                if from_string(line.date_to) <= from_string(line.paid_date):
+                    theo_amt = 0.00
+                else:
+                    theo_amt = line.planned_amount
+            else:
+                line_timedelta = from_string(line.date_to) - from_string(line.date_from)
+                elapsed_timedelta = from_string(today) - (from_string(line.date_from))
```
Comment (astirpe) [inline account_budget_oca/models/account_budget.py]:
```suggestion
                elapsed_timedelta = from_string(today) - from_string(line.date_from)
```

File: account_budget_oca/models/account_budget.py [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+                    theo_amt = (
+                        elapsed_timedelta.total_seconds()
+                        / line_timedelta.total_seconds()
+                    ) * line.planned_amount
+                else:
+                    theo_amt = line.planned_amount
+
+            line.theoretical_amount = theo_amt
+
+    def _compute_percentage(self):
```
Comment (astirpe) [inline account_budget_oca/models/account_budget.py:209]:
```suggestion
    @api.depends('theoretical_amount', 'practical_amount')
    def _compute_percentage(self):
```

File: account_budget_oca/tests/common.py [hunk 19]
```diff
@@ -0,0 +1,271 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import time
+
+from odoo.addons.account.tests.account_test_classes import AccountingTestCase
+
+
+class TestAccountBudgetCommon(AccountingTestCase):
+    def setUp(self):
+        super(TestAccountBudgetCommon, self).setUp()
```
Comment (astirpe) [inline account_budget_oca/tests/common.py]:
```suggestion
        super().setUp()
```

File: account_budget_oca/tests/test_theoreticalamount.py [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+except ImportError:
+    from mock import patch
+
+
+# ---------------------------------------------------------
+# Tests
+# ---------------------------------------------------------
+class TestTheoreticalAmount(TestAccountBudgetCommon):
+    def setUp(self):
+        super(TestTheoreticalAmount, self).setUp()
```
Comment (astirpe) [inline account_budget_oca/tests/test_theoreticalamount.py]:
```suggestion
        super().setUp()
```

File: account_budget_oca/tests/test_theoreticalamount.py [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+
+    def test_10(self):
+        """At last"""
+        date = Datetime.to_string(Datetime.from_string("2014-12-31 00:00:00"))
+        self.mock_datetime.now.return_value = date
+        self.assertAlmostEqual(self.line.theoretical_amount, -364)
+
+    def tearDown(self):
+        self.patcher.stop()
+        super(TestTheoreticalAmount, self).tearDown()
```
Comment (astirpe) [inline account_budget_oca/tests/test_theoreticalamount.py]:
```suggestion
        super().tearDown()
```

File: account_budget_oca/tests/test_theoreticalamount.py [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+                "crossovered_budget_id": crossovered_budget.id,
+                "general_budget_id": budget_post.id,
+                "date_from": "2014-01-01",
+                "date_to": "2014-12-31",
+                "planned_amount": -364,
+            }
+        )
+
+        self.patcher = patch(
+            "odoo.addons.account_budget_oca.models." "account_budget.fields.Datetime",
```
Comment (astirpe) [inline account_budget_oca/tests/test_theoreticalamount.py]:
```suggestion
            "odoo.addons.account_budget_oca.models.account_budget.fields.Datetime",
```

File: account_budget_oca/views/account_analytic_account_views.xml [hunk 23]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record model="ir.ui.view" id="view_account_analytic_account_form_inherit_budget">
```
Comment (astirpe) [inline account_budget_oca/views/account_analytic_account_views.xml]:
```suggestion
    <record id="view_account_analytic_account_form_inherit_budget" model="ir.ui.view" >
```

File: account_budget_oca/views/account_analytic_account_views.xml [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+                <notebook groups="account.group_account_user">
+                    <page string="Budget Items">
+                        <field
+                            name="crossovered_budget_line_ids"
+                            widget="one2many_list"
+                            colspan="4"
+                            nolabel="1"
+                            mode="tree"
+                        >
+                            <tree string="Budget Items" editable="top">
```
Comment (astirpe) [inline account_budget_oca/views/account_analytic_account_views.xml]:
```suggestion
                            <tree editable="top">
```

File: account_budget_oca/views/account_analytic_account_views.xml [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+                                    widget="monetary"
+                                />
+                                <field
+                                    name="theoretical_amount"
+                                    sum="Theoretical Amount"
+                                    widget="monetary"
+                                />
+                                <field name="percentage" />
+                            </tree>
+                            <form string="Budget Items">
```
Comment (astirpe) [inline account_budget_oca/views/account_analytic_account_views.xml]:
```suggestion
                            <form>
```

File: account_budget_oca/views/account_budget_views.xml [hunk 26]
```diff
@@ -0,0 +1,464 @@
...
+                />
+                <field name="company_id" groups="base.group_multi_company" />
+            </search>
+        </field>
+    </record>
+    <record id="view_budget_post_tree" model="ir.ui.view">
+        <field name="name">account.budget.post.tree</field>
+        <field name="model">account.budget.post</field>
+        <field name="arch" type="xml">
+            <tree string="Budgetary Position">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
            <tree>
```

File: account_budget_oca/views/account_budget_views.xml [hunk 27]
```diff
@@ -0,0 +1,464 @@
...
+                    <notebook>
+                        <page string="Budget Lines">
+                            <field
+                                name="crossovered_budget_line_ids"
+                                context="{'default_date_from': date_from,'default_date_to': date_to}"
+                                colspan="4"
+                                nolabel="1"
+                                attrs="{'readonly':[('state','!=','draft')]}"
+                            >
+                                <tree string="Budget Lines">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
                                <tree>
```

File: account_budget_oca/views/account_budget_views.xml [hunk 28]
```diff
@@ -0,0 +1,464 @@
...
+            <search string="Budget Lines">
+                <field name="analytic_account_id" />
+            </search>
+        </field>
+    </record>
+    <record model="ir.ui.view" id="view_crossovered_budget_line_tree">
+        <field name="name">crossovered.budget.line.tree</field>
+        <field name="model">crossovered.budget.lines</field>
+        <field name="arch" type="xml">
+            <tree string="Budget Lines">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
            <tree>
```

File: account_budget_oca/views/account_budget_views.xml [hunk 29]
```diff
@@ -0,0 +1,464 @@
...
+        action="open_budget_post_form"
+        id="menu_budget_post_form"
+        parent="account.account_management_menu"
+        sequence="5"
+    />
+    <record model="ir.ui.view" id="view_budget_post_form">
+        <field name="name">account.budget.post.form</field>
+        <field name="model">account.budget.post</field>
+        <field name="arch" type="xml">
+            <form string="Budgetary Position">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
            <form>
```

File: account_budget_oca/views/account_budget_views.xml [hunk 30]
```diff
@@ -0,0 +1,464 @@
...
+                    </page>
+                </notebook>
+            </form>
+        </field>
+    </record>
+    <record model="ir.ui.view" id="crossovered_budget_view_form">
+        <field name="name">crossovered.budget.view.form</field>
+        <field name="model">crossovered.budget</field>
+        <field name="arch" type="xml">
+            <form string="Budget">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
            <form>
```

File: account_budget_oca/views/account_budget_views.xml [hunk 31]
```diff
@@ -0,0 +1,464 @@
...
+                                        widget="monetary"
+                                    />
+                                    <field
+                                        name="theoretical_amount"
+                                        sum="Theoretical Amount"
+                                        widget="monetary"
+                                    />
+                                    <field name="percentage" />
+                                </tree>
+                                <form string="Budget Lines">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
                                <form>
```

File: account_budget_oca/views/account_budget_views.xml [hunk 32]
```diff
@@ -0,0 +1,464 @@
...
+                <field name="theoretical_amount" widget="monetary" />
+                <field name="percentage" />
+            </tree>
+        </field>
+    </record>
+    <record model="ir.ui.view" id="view_crossovered_budget_line_form">
+        <field name="name">crossovered.budget.line.form</field>
+        <field name="model">crossovered.budget.lines</field>
+        <field name="arch" type="xml">
+            <form string="Budget Lines">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
            <form>
```

File: account_budget_oca/views/account_budget_views.xml [hunk 33]
```diff
@@ -0,0 +1,464 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <!--
+    Budgets
+    -->
+    <record id="view_budget_post_search" model="ir.ui.view">
+        <field name="name">account.budget.post.search</field>
+        <field name="model">account.budget.post</field>
+        <field name="arch" type="xml">
+            <search string="Budgetary Position">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
            <search>
```

File: account_budget_oca/views/account_budget_views.xml [hunk 34]
```diff
@@ -0,0 +1,464 @@
...
+                    </t>
+                </templates>
+            </kanban>
+        </field>
+    </record>
+    <record id="view_crossovered_budget_search" model="ir.ui.view">
+        <field name="name">crossovered.budget.search</field>
+        <field name="model">crossovered.budget</field>
+        <field name="arch" type="xml">
+            <search string="Budget">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
            <search>
```

File: account_budget_oca/views/account_budget_views.xml [hunk 35]
```diff
@@ -0,0 +1,464 @@
...
+        name="Budgets"
+        action="act_crossovered_budget_view"
+        sequence="60"
+        groups="account.group_account_user"
+    />
+    <record id="view_crossovered_budget_line_search" model="ir.ui.view">
+        <field name="name">account.budget.line.search</field>
+        <field name="model">crossovered.budget.lines</field>
+        <field name="arch" type="xml">
+            <search string="Budget Lines">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
            <search>
```

File: account_budget_oca/views/account_budget_views.xml [hunk 36]
```diff
@@ -0,0 +1,464 @@
...
+        </field>
+    </record>
+    <record model="ir.ui.view" id="crossovered_budget_view_tree">
+        <field name="name">crossovered.budget.view.tree</field>
+        <field name="model">crossovered.budget</field>
+        <field name="arch" type="xml">
+            <tree
+                decoration-info="state == 'draft'"
+                decoration-muted="state in ('done','cancel')"
+                string="Budget"
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml:257]:
```suggestion
```

File: account_budget_oca/views/account_budget_views.xml [hunk 37]
```diff
@@ -0,0 +1,464 @@
...
+                        name="message_follower_ids"
+                        widget="mail_followers"
+                        groups="base.group_user"
+                    />
+                    <field name="message_ids" widget="mail_thread" />
+                </div>
+            </form>
+        </field>
+    </record>
+    <record model="ir.ui.view" id="crossovered_budget_view_tree">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
    <record id="crossovered_budget_view_tree" model="ir.ui.view">
```

File: account_budget_oca/views/account_budget_views.xml [hunk 38]
```diff
@@ -0,0 +1,464 @@
...
+                    string="To Approve"
+                    name="to_approve"
+                    domain="[('state','=','confirm')]"
+                    help="To Approve Budgets"
+                />
+                <field name="state" />
+            </search>
+        </field>
+    </record>
+    <record model="ir.actions.act_window" id="act_crossovered_budget_view">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
    <record id="act_crossovered_budget_view" model="ir.actions.act_window">
```

File: account_budget_oca/views/account_budget_views.xml [hunk 39]
```diff
@@ -0,0 +1,464 @@
...
+    <record id="view_crossovered_budget_line_search" model="ir.ui.view">
+        <field name="name">account.budget.line.search</field>
+        <field name="model">crossovered.budget.lines</field>
+        <field name="arch" type="xml">
+            <search string="Budget Lines">
+                <field name="analytic_account_id" />
+            </search>
+        </field>
+    </record>
+    <record model="ir.ui.view" id="view_crossovered_budget_line_tree">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
    <record id="view_crossovered_budget_line_tree" model="ir.ui.view">
```

File: account_budget_oca/views/account_budget_views.xml [hunk 40]
```diff
@@ -0,0 +1,464 @@
...
+                <field name="date_to" />
+                <field name="paid_date" groups="base.group_no_one" />
+                <field name="planned_amount" widget="monetary" />
+                <field name="practical_amount" widget="monetary" />
+                <field name="theoretical_amount" widget="monetary" />
+                <field name="percentage" />
+            </tree>
+        </field>
+    </record>
+    <record model="ir.ui.view" id="view_crossovered_budget_line_form">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
    <record id="view_crossovered_budget_line_form" model="ir.ui.view">
```

File: account_budget_oca/views/account_budget_views.xml [hunk 41]
```diff
@@ -0,0 +1,464 @@
...
+                            name="company_id"
+                            options="{'no_create': True}"
+                            groups="base.group_multi_company"
+                        />
+                    </group>
+                </sheet>
+            </form>
+        </field>
+    </record>
+    <record model="ir.actions.act_window" id="act_crossovered_budget_lines_view">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
    <record id="act_crossovered_budget_lines_view" model="ir.actions.act_window">
```

File: account_budget_oca/views/account_budget_views.xml [hunk 42]
```diff
@@ -0,0 +1,464 @@
...
+                                <field name="code" />
+                                <field name="name" />
+                            </tree>
+                        </field>
+                    </page>
+                </notebook>
+            </form>
+        </field>
+    </record>
+    <record model="ir.ui.view" id="crossovered_budget_view_form">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
    <record id="crossovered_budget_view_form" model="ir.ui.view">
```

File: account_budget_oca/views/account_budget_views.xml [hunk 43]
```diff
@@ -0,0 +1,464 @@
...
+        <field name="view_id" ref="view_budget_post_tree" />
+        <field name="search_view_id" ref="view_budget_post_search" />
+    </record>
+    <menuitem
+        action="open_budget_post_form"
+        id="menu_budget_post_form"
+        parent="account.account_management_menu"
+        sequence="5"
+    />
+    <record model="ir.ui.view" id="view_budget_post_form">
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]:
```suggestion
    <record id="view_budget_post_form" model="ir.ui.view">
```

File: account_budget_oca/security/account_budget_security.xml [hunk 44]
```diff
@@ -0,0 +1,35 @@
...
+        </record>
+        <record model="res.users" id="base.user_root">
```
Comment (astirpe) [inline account_budget_oca/security/account_budget_security.xml]:
```suggestion
        <record id="base.user_root" model="res.users">
```

File: account_budget_oca/views/account_budget_views.xml [hunk 45]
```diff
@@ -0,0 +1,464 @@
...
+        sequence="20"
+    />
+    <!--  Shortcuts -->
+    <act_window
+        name="Budget Items"
+        context="{'search_default_analytic_account_id': [active_id], 'default_analytic_account_id': active_id}"
+        res_model="crossovered.budget.lines"
+        binding_model="account.analytic.account"
+        id="act_account_analytic_account_cb_lines"
+    />
```
Comment (astirpe) [inline account_budget_oca/views/account_budget_views.xml]: This window action could be removed

**Review by astirpe — CHANGES_REQUESTED** [review]
Only code review. Not tested.

**Comment by amcor** [comment]
@astirpe thanks for the review !! 
I think  i've done all suggestions.

File: account_budget_oca/security/account_budget_security.xml [hunk 46]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <data noupdate="1">
+        <record id="budget_post_comp_rule" model="ir.rule">
+            <field name="name">Budget post multi-company</field>
+            <field name="model_id" ref="model_account_budget_post" />
+            <field eval="True" name="global" />
+            <field
+                name="domain_force"
+            >['|',('company_id','=',False),('company_id','in',company_ids))]</field>
```
Comment (astirpe) [inline account_budget_oca/security/account_budget_security.xml]:
```suggestion
            >['|',('company_id','=',False),('company_id','in',company_ids)]</field>
```

File: account_budget_oca/security/account_budget_security.xml [hunk 47]
```diff
@@ -0,0 +1,35 @@
...
+        </record>
+        <record id="budget_comp_rule" model="ir.rule">
+            <field name="name">Budget multi-company</field>
+            <field name="model_id" ref="model_crossovered_budget" />
+            <field eval="True" name="global" />
+            <field
+                name="domain_force"
+            >['|',('company_id','=',False),('company_id','in',company_ids))]</field>
```
Comment (astirpe) [inline account_budget_oca/security/account_budget_security.xml]:
```suggestion
            >['|',('company_id','=',False),('company_id','in',company_ids)]</field>
```

File: account_budget_oca/security/account_budget_security.xml [hunk 48]
```diff
@@ -0,0 +1,35 @@
...
+        </record>
+        <record id="budget_lines_comp_rule" model="ir.rule">
+            <field name="name">Budget lines multi-company</field>
+            <field name="model_id" ref="model_crossovered_budget_lines" />
+            <field eval="True" name="global" />
+            <field
+                name="domain_force"
+            >['|',('company_id','=',False),('company_id','in',company_ids))]</field>
```
Comment (astirpe) [inline account_budget_oca/security/account_budget_security.xml]:
```suggestion
            >['|',('company_id','=',False),('company_id','in',company_ids)]</field>
```

File: account_budget_oca/views/account_budget_views.xml (hunk 37 above)
Comment (astirpe) [reply account_budget_oca/views/account_budget_views.xml]: Still pending

File: account_budget_oca/views/account_budget_views.xml (hunk 41 above)
Comment (astirpe) [reply account_budget_oca/views/account_budget_views.xml]: still pending

File: account_budget_oca/views/account_budget_views.xml (hunk 32 above)
Comment (astirpe) [reply account_budget_oca/views/account_budget_views.xml]: still pending

File: account_budget_oca/views/account_budget_views.xml (hunk 33 above)
Comment (astirpe) [reply account_budget_oca/views/account_budget_views.xml]: still pending

File: account_budget_oca/views/account_budget_views.xml (hunk 26 above)
Comment (astirpe) [reply account_budget_oca/views/account_budget_views.xml]: still pending

File: account_budget_oca/views/account_budget_views.xml (hunk 29 above)
Comment (astirpe) [reply account_budget_oca/views/account_budget_views.xml]: still pending

File: account_budget_oca/views/account_budget_views.xml (hunk 30 above)
Comment (astirpe) [reply account_budget_oca/views/account_budget_views.xml]: still pending

File: account_budget_oca/views/account_budget_views.xml (hunk 27 above)
Comment (astirpe) [reply account_budget_oca/views/account_budget_views.xml]: still pending

File: account_budget_oca/views/account_budget_views.xml (hunk 31 above)
Comment (astirpe) [reply account_budget_oca/views/account_budget_views.xml]: still pending

File: account_budget_oca/views/account_budget_views.xml (hunk 36 above)
Comment (astirpe) [reply account_budget_oca/views/account_budget_views.xml:257]: still pending

File: account_budget_oca/views/account_budget_views.xml (hunk 34 above)
Comment (astirpe) [reply account_budget_oca/views/account_budget_views.xml]: still pending

File: account_budget_oca/views/account_budget_views.xml (hunk 35 above)
Comment (astirpe) [reply account_budget_oca/views/account_budget_views.xml]: still pending

File: account_budget_oca/views/account_budget_views.xml (hunk 28 above)
Comment (astirpe) [reply account_budget_oca/views/account_budget_views.xml]: still pending

**Review by astirpe — CHANGES_REQUESTED** [review]
Please add `migrations/13.0.1.0.0/post-migrate.py` file with migration script, something like:


```python
# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html)

from odoo import SUPERUSER_ID, api


def migrate(cr, version):
    env = api.Environment(cr, SUPERUSER_ID, {})
    domain = "['|',('company_id','=',False),('company_id','in',company_ids)]"
    rule = env.ref("account_budget_oca.budget_post_comp_rule", raise_if_not_found=False)
    if rule:
        rule.write({"domain_force": domain})
    rule = env.ref("account_budget_oca.budget_comp_rule", raise_if_not_found=False)
    if rule:
        rule.write({"domain_force": domain})

    rule = env.ref("account_budget_oca.budget_lines_comp_rule", raise_if_not_found=False)
    if rule:
        rule.write({"domain_force": domain})

```

**Review by astirpe — APPROVED** [review]
Code review only (not tested)

**Comment by RachidAlassir** [comment]
@amcor 
In Accounting -> Management -> budget: After clicking at "Add line" (in Budget lines)
This error appears:
```
  File "/home/.../account/account_budget_oca/models/account_budget.py", line 186, in _compute_theoretical_amount
    line_timedelta = from_string(line.date_to) - from_string(line.date_from)
TypeError: unsupported operand type(s) for -: 'NoneType' and 'NoneType'
```

**Comment by flachica** [comment]
> @amcor
> In Accounting -> Management -> budget: After clicking at "Add line" (in Budget lines)
> This error appears:
> 
> ```
>   File "/home/.../account/account_budget_oca/models/account_budget.py", line 186, in _compute_theoretical_amount
>     line_timedelta = from_string(line.date_to) - from_string(line.date_from)
> TypeError: unsupported operand type(s) for -: 'NoneType' and 'NoneType'
> ```
I have just performed the operations you describe and the problem does not appear. The dates on the lines are now required. I don't know if when you tested the module they were

File: account_budget_oca/models/account_budget.py [hunk 49]
```diff
@@ -0,0 +1,216 @@
...
+        today = fields.Datetime.now()
+        for line in self:
+            # Used for the report
+
+            if line.paid_date:
+                if from_string(line.date_to) <= from_string(line.paid_date):
+                    theo_amt = 0.00
+                else:
+                    theo_amt = line.planned_amount
+            else:
```
Comment (gaikaz) [inline account_budget_oca/models/account_budget.py]:
```suggestion
            elif line.date_from and line.date_to:
```

part (1/2)

File: account_budget_oca/models/account_budget.py [hunk 50]
```diff
@@ -0,0 +1,216 @@
...
+                    # If today is between the budget line date_from and
+                    # date_to
+                    theo_amt = (
+                        elapsed_timedelta.total_seconds()
+                        / line_timedelta.total_seconds()
+                    ) * line.planned_amount
+                else:
+                    theo_amt = line.planned_amount
+
+            line.theoretical_amount = theo_amt
```
Comment (gaikaz) [inline account_budget_oca/models/account_budget.py]:
```suggestion
            else:
                theo_amt = 0.00
            line.theoretical_amount = theo_amt
```
part (2/2)

**Review by gaikaz — CHANGES_REQUESTED** [review]
To fix the issue @RachidAlassir mentioned in [his comment](https://github.com/OCA/account-budgeting/pull/37#issuecomment-642356914)
My colleague also stumbled upon it today with the newest version while testing the module (so with the required fields, @flachica)

File: account_budget_oca/models/account_budget.py (hunk 50 above)
Comment (amcor) [reply account_budget_oca/models/account_budget.py]: sorry @RachidAlassir, i left this totally unattended .. and forgotten :(
thanks @gaikaz for the suggestions ! they are already applied !

**Review by gaikaz — APPROVED** [review]
A client has been testing this for their use since March.
Must be functionally solid.

**Comment by unique4** [comment]
All check passed, why is this PR not merged?

**Pull request merged**